### PR TITLE
Self describing

### LINKED ISSUE #2: Decoding advanced serde constructions like tagged enums
Ok, after the #1, I digged Serde sources, and one comment says that advanced features like tagged enums require "self-descriptive" formats. While bencode seems to be self-descriptive, it doesn't distinguish byte string and Unicode strings, and Serde expects Unicode strings.

There is a simple solution: the `deserialize_any` method could check if byte sequence is a valid UTF-8 string, and call either "visit_str" or "visit_bytes" depending on the result.  However, for those who do not need these advanced features, it will impose CPU penalty.  We might have two implementations (`Deserializer` and `AutoDeserializer`), or simply have a feature flag.  First possibility is more flexible, and second is easier to implement and maintain. What do you think?

The deserialize_any diff looks like:
```diff
-            b'0'..=b'9' => self.deserialize_str(visitor),
+            b'0'..=b'9' => {
+                let bytes = self.parse_byte_string()?;
+                match std::str::from_utf8(bytes) {
+                    Ok(s) => visitor.visit_borrowed_str(s),
+                    Err(_) => visitor.visit_borrowed_bytes(bytes),
+                }
+            }
```

### PR DESCRIPTION
**Pull request by knightpp**
Fixes #2

### CODE REVIEW

**Comment by knightpp** [comment]
@monoid Hey, what do you think?

File: src/de.rs [hunk 1]
```diff
@@ -582,8 +669,73 @@ mod tests {
...
+            fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                deserializer.deserialize_any(TestAutoVisitor)
+            }
+        }
+        let input = b"d6:string4:12345:bytes4:"
+            .iter()
+            .chain([1, 2, 3, 4].iter())
```
Comment (monoid) [inline src/de.rs]: The string can be written as `b"d6:string4:12345:bytes4:\x01\x02\x03\x04e"`.

File: src/de.rs [hunk 2]
```diff
@@ -56,16 +136,19 @@ pub fn from_bytes<'a, T>(b: &'a [u8]) -> Result<T>
...
-    if deserializer.input.is_empty() {
-        Ok(t)
-    } else {
-        Err(Error::SyntaxError(deserializer.input[0], None))
-    }
+    _from_slice::<T, Simple>(b)
+}
+/// The same as [`from_bytes`] but `deserialize_any`
+/// will deserialize byte string as `str` if
+/// input bytes are valid UTF-8, otherwise as `bytes`
```
Comment (monoid) [inline src/de.rs]: ... thus making ipossible to deserialize advanced Serde constructs like tagged enums with some penalty for simpler use cases.

**Review by monoid — COMMENTED** [review]
Cool!

It would be nice to have "normal" tests to run with both auto and simple deserializers, though.

**Comment by knightpp** [comment]
Thanks!

**Comment by monoid** [comment]
Thank you!  I forgot to mention that my app's tests that uses different types of enums, pass with this branch!

File: src/de.rs [hunk 3]
```diff
@@ -582,8 +677,68 @@ mod tests {
...
+
+        impl<'de> de::Deserialize<'de> for TestAuto {
+            fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                deserializer.deserialize_any(TestAutoVisitor)
+            }
+        };
+        let input = b"d6:string4:12345:bytes4:\x01\x02\x03\x04e";
```
Comment (monoid) [inline src/de.rs:733]: BTW, `b"\x01\x02\x03\x04"` is a valid UTF-8.  Just put `\xFF` at any place to make it invalid UTF-8 and make the test stricter.

Comment (knightpp) [reply src/de.rs:733]: Hm, I thought control characters are invalid UTF-8. Great catch!

Comment (monoid) [reply src/de.rs:733]: You are welcome!

UTF-8 is backward compatible with `0x00..=0xFF` ASCII range.

Comment (monoid) [reply src/de.rs:733]: ... `0x00..=0x7F` ASCII range.

**Pull request merged**